src/Features/Core/Portable/Extensions/ExtensionMessageHandlerService.cs

```diff
@@ -0,0 +1,358 @@
+            }
+        }
+
+        extension?.AnalyzerAssemblyLoader.Dispose();
+
+        return ValueTask.CompletedTask;
+    }
+
+    public void Reset()
+    {
+        List<Extension> extensions;
+        lock (_lockObject)
+        {
+            extensions = _extensions.Values.ToList();
+            _extensions.Clear();
+            _workspaceHandlers.Clear();
+            _documentHandlers.Clear();
+        }
+
+        foreach (var extension in extensions)
+        {
+            extension.AnalyzerAssemblyLoader.Dispose();
+        }
+    }
+
+    public void Dispose()
+        => Reset();
+
+    private async ValueTask<string> HandleExtensionMessageAsync<TArgument>(
+        TArgument argument,
+        string messageName,
+        string jsonMessage,
+        Dictionary<string, IExtensionMessageHandlerWrapper<TArgument>> handlers,
+        CancellationToken cancellationToken)
+    {
+        IExtensionMessageHandlerWrapper<TArgument> handler;
+        lock (_lockObject)
+        {
+            // handlers here is either _workspaceHandlers or _documentHandlers, so it must be protected
+            // by _lockObject.
+            if (!handlers.TryGetValue(messageName, out handler!))
+            {
+                throw new InvalidOperationException($"No handler found for message {messageName}.");
+            }
+        }
+
+        // Any exception thrown in this method is left to bubble up to the extension.
+        var message = JsonSerializer.Deserialize(jsonMessage, handler.MessageType);
+        var result = await handler.ExecuteAsync(message, argument, cancellationToken)
+            .ConfigureAwait(false);
+        var responseJson = JsonSerializer.Serialize(result, handler.ResponseType);
+        return responseJson;
+    }
+
+    private void RegisterAssembly(Extension extension, string assemblyFileName, AssemblyHandlers? assemblyHandlers)
+    {
+        lock (_lockObject)
+        {
+            // Make sure a call to UnloadCustomMessageHandlersAsync hasn't happened while we relinquished the lock on _lockObject
+            if (!_extensions.TryGetValue(extension.AssemblyFolderPath, out var currentExtension) || !currentExtension.Equals(extension))
+            {
+                throw new InvalidOperationException($"Handlers in {extension.AssemblyFolderPath} were unregistered while loading handlers.");
+            }
+
+            try
+            {
+                if (assemblyHandlers is not null)
+                {
+                    var duplicateHandler = _workspaceHandlers.Keys.Intersect(assemblyHandlers.WorkspaceMessageHandlers.Keys).Concat(
+                    _documentHandlers.Keys.Intersect(assemblyHandlers.DocumentMessageHandlers.Keys)).FirstOrDefault();
+
+                    if (duplicateHandler is not null)
+                    {
+                        assemblyHandlers = null;
+                        throw new InvalidOperationException($"Handler name {duplicateHandler} is already registered.");
```

Comment:
same issue with IOE.  these should only be for bugs in OUR code, not misconfgured extensions or users doing something innapropriate.  those should be detected and gracefully logged/handled.